contracts/validator-manager/SlotAuctionManager.sol

```diff
@@ -5,242 +5,440 @@
+            ) {
+                revert MoreActiveValidatorsThanTotalSlots(
+                    auctionSettings.totalValidatorSlots,
+                    $._totalValidatorSlots - $._openValidatorSlots
+                );
+            }
+            $._openValidatorSlots -= $._totalValidatorSlots - auctionSettings.totalValidatorSlots;
+            $._totalValidatorSlots = auctionSettings.totalValidatorSlots;
+        }
+        $._minAuctionDuration = auctionSettings.minAuctionDuration;
+        $._minValidatorDuration = auctionSettings.minValidatorDuration;
+        $._minimumBid = auctionSettings.minimumBid;
+        $._validatorWeight = auctionSettings.weight;
+    }
+
+    function minBidRequired() external view AuctionOn returns (uint256) {
+        SlotAuctionManagerStorage storage $ = _getSlotAuctionManagerStorage();
+
+        if (Heap.length($._bids) < $._totalValidatorSlots) {
+            return $._minimumBid;
         }
-        return Heap.peek(_bids) + 1;
+        return Heap.peek($._bids) + 1;
     }
 
-    function _placeBid (
+    function _placeBid(
         uint256 bid,
         bytes memory nodeID,
         bytes memory blsPublicKey,
         PChainOwner memory remainingBalanceOwner,
         PChainOwner memory disableOwner
-    ) nonReentrant AuctionOn internal {
-        if (VALIDATOR_MANAGER.getNodeValidationID(nodeID) != 0) {
+    ) internal AuctionOn {
+        SlotAuctionManagerStorage storage $ = _getSlotAuctionManagerStorage();
+
+        if ($._manager.getNodeValidationID(nodeID) != 0) {
             revert NodeIsValidator(nodeID);
         }
-        if (_nodeIsQualified[nodeID]) {
-            revert NodeIsWinning(nodeID);
+        if ($._nodeIsQualified[nodeID]) {
+            revert DuplicateNodeIDInContention(nodeID);
         }
-        if (bidderInfo[bid].addr != address(0)) {
+        if ($._bidderInfo[bid].addr != address(0)) {
             revert DuplicateBidInContention(bid);
         }
-        if (MinimumBid > bid) {
-            revert BidSmallerThanMinimum(MinimumBid, bid);
+        if ($._minimumBid > bid) {
+            revert BidSmallerThanMinimum($._minimumBid, bid);
         }
 
         // If all slots aren't contended, then fill the heap with any bid
-        if (Heap.length(_bids) < validatorSlots) {
-            TOKEN_CONTRACT.transferFrom(msg.sender, address(this), bid);
-            Heap.insert(_bids, bid);
-        } 
-        else if (Heap.peek(_bids) < bid) {
+        if (Heap.length($._bids) < $._auctioningValidatorSlots) {
+            _lock(bid);
+            Heap.insert($._bids, bid);
+        } else if (Heap.peek($._bids) < bid) {
             _lock(bid);
-            // TOKEN_CONTRACT.transferFrom(msg.sender, address(this), bid);
-            uint256 poppedBid = Heap.replace(_bids, bid);
-            _secondPrice = poppedBid;
-            emit BidEvicted(poppedBid, bidderInfo[poppedBid].nodeID);
+
+            $._secondPrice = Heap.replace($._bids, bid);
```

Comment:
imo we should store this in a variable called `evictedBid` and then assign it to `$._secondPrice`. Right now calling `_unlock` on `_secondPrice` is weird and slightly confusing